lib/grandpa/grandpa.go

```diff
@@ -1297,3 +1209,192 @@ func (s *Service) lenVotes(stage Subround) int {
+	}
+
+	s.telemetry.SendMessage(
+		telemetry.NewAfgReceivedCommit(
+			msg.Vote.Hash,
+			fmt.Sprint(msg.Vote.Number),
+			containsPrecommitsSignedBy,
+		),
+	)
+
+	if has, _ := s.blockState.HasFinalisedBlock(msg.Round, s.state.setID); has {
+		return nil
+	}
+
+	// check justification here
+	err = verifyCommitMessageJustification(msg, s.state.setID,
+		s.state.threshold(), s.authorities(), s.blockState)
+	if err != nil {
+		if errors.Is(err, blocktree.ErrStartNodeNotFound) {
+			// we haven't synced the committed block yet, add this to the tracker for later processing
+			s.tracker.addCommit(msg)
+		}
+		return err
+	}
+
+	err = s.blockState.SetFinalisedHash(msg.Vote.Hash, msg.Round, s.state.setID)
+	if err != nil {
+		return fmt.Errorf("setting finalised hash %s: %w", msg.Vote.Hash.Short(), err)
+	}
+
+	pcs, err := compactToJustification(msg.Precommits, msg.AuthData)
+	if err != nil {
+		return fmt.Errorf("compacting justification: %w", err)
+	}
+
+	err = s.grandpaState.SetPrecommits(msg.Round, msg.SetID, pcs)
+	if err != nil {
+		return fmt.Errorf("setting precommits: %w", err)
+	}
+
+	// ensure the persistence of informations
+	// and then notify the grandpa service about
+	// a received commit message
+	s.receivedCommit <- msg
+
+	// TODO: re-add catch-up logic (#1531)
+	return nil
+}
+
+func verifyCommitMessageJustification(fm *CommitMessage, setID uint64, threshold uint64,
+	authorities []*types.Authority, blockState BlockState) error {
+	if len(fm.Precommits) != len(fm.AuthData) {
+		return ErrPrecommitSignatureMismatch
+	}
+
+	if fm.SetID != setID {
+		return fmt.Errorf("%w: grandpa state set id %d, set id in the commit message %d",
+			ErrSetIDMismatch, setID, fm.SetID)
+	}
+
+	highestFinalizedHeader, err := blockState.GetHighestFinalisedHeader()
+	if err != nil {
+		return fmt.Errorf("getting highest finalised header: %w", err)
+	}
+
+	isDescendant, err := blockState.IsDescendantOf(highestFinalizedHeader.Hash(), fm.Vote.Hash)
+	if err != nil {
+		return fmt.Errorf("cannot verify ancestry of highest finalised block: %w", err)
+	}
+
+	if !isDescendant {
+		return errVoteBlockMismatch
+	}
+
+	eqvVoters := getEquivocatoryVoters(fm.AuthData)
+	var count int
+	for i, pc := range fm.Precommits {
+		just := &SignedVote{
+			Vote:        pc,
+			Signature:   fm.AuthData[i].Signature,
+			AuthorityID: fm.AuthData[i].AuthorityID,
+		}
+
+		err := verifyJustification(just, fm.Round, setID, precommit, authorities)
+		if err != nil {
+			logger.Errorf("failed to verify justification for vote from authority id %s, for block hash %s: %s",
+				just.AuthorityID.String(), just.Vote.Hash, err)
+			continue
+		}
+
+		isDescendant, err := blockState.IsDescendantOf(fm.Vote.Hash, just.Vote.Hash)
+		if err != nil {
+			logger.Warnf("could not check for descendant: %s", err)
+			continue
+		}
+
+		precommitedHeader, err := blockState.GetHeader(pc.Hash)
+		if err != nil {
+			return fmt.Errorf("getting header of %s: %w", pc.Hash.Short(), err)
```

Comment:
the `pc.Hash` information should be wrapped inside `GetHeader`, not here

```suggestion
			return fmt.Errorf("getting header: %w", err)
```